integration/integration_test_base.py

```diff
@@ -0,0 +1,157 @@
+"""@package integration_test_base
+Base classes for Nighthawk integration tests
+"""
+
+import json
+import logging
+from nighthawk_test_server import NighthawkTestServer
+import os
+import threading
+import time
+import subprocess
+import sys
+import socket
+import unittest
+
+
+class IntegrationTestBase(unittest.TestCase):
+    """
+    IntegrationTestBase facilitates testing against the Nighthawk test server, by running
+    determining a free port, and starting it up in a separate process in setUp().
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(IntegrationTestBase, self).__init__(*args, **kwargs)
+        self.test_rundir = os.path.join(os.environ["TEST_SRCDIR"],
+                                        os.environ["TEST_WORKSPACE"])
+
+        self.nighthawk_test_server_path = os.path.join(
+            self.test_rundir, "nighthawk_test_server")
+        self.nighthawk_test_config_path = None
+        self.nighthawk_client_path = os.path.join(
+            self.test_rundir, "nighthawk_client")
+        self.server_ip = "::1" if IntegrationTestBase.ipv6 else "127.0.0.1"
+        self.socket_type = socket.AF_INET6 if IntegrationTestBase.ipv6 else socket.AF_INET
+        self.test_server = None
+        self.server_port = -1
+        self.parameters = dict()
+
+    def getFreeListenerPortForAddress(self, address):
+        """
+        Determines a free port and returns that. Theoretically it is possible that another process
```

Comment:
The clean way to do this is to start the server bound to port 0, then query its `/listeners` handle on the admin port. Then you need to know what the admin port is, which is ideally also bound to 0 in the config. To figure that out, you can do something like https://github.com/envoyproxy/envoy/blob/master/test/integration/hotrestart_test.sh#L72.

Reply:
I agree that approach is cleaner. But, I think that soon(ish) we may also want to be running other proxies and services for running comparison test runs, which is what led me up to this approach.
Because when we want to add other server types, not having to implement the knowledge about how to get to the correct port(s) for each of them is a win? 
E.g. adding support for nginx in the current approach is easy (see (see `NighthawkTestServer(TestServerBase)`), whereas I wouldn't know how to query it for a port that got assigned for a specific listener.

Is the concern that the port we intended for a server to listen on gets bound to by another process else before our server gets around to that?

Reply:
Will add an item for tracking we can punt this to a follow up